File: .sdk-repos/version-8.9/orchestration-cluster-api-csharp/examples/ProcessInstance.cs
// Compilable usage examples for process instance operations.
// These examples are type-checked during build to guard against API regressions.
using Camunda.Orchestration.Sdk;

public static class ProcessInstanceExamples
{
    #region CreateProcessInstanceById
    // <CreateProcessInstanceById>
    public static async Task CreateProcessInstanceByIdExample(ProcessDefinitionId processDefinitionId)
    {
        using var client = CamundaClient.Create();

        var result = await client.CreateProcessInstanceAsync(new ProcessInstanceCreationInstructionById
        {
            ProcessDefinitionId = processDefinitionId,
        });

        Console.WriteLine($"Process instance key: {result.ProcessInstanceKey}");
    }
    // </CreateProcessInstanceById>
    #endregion CreateProcessInstanceById

    #region CreateProcessInstanceByKey

    // <CreateProcessInstanceByKey>
    public static async Task CreateProcessInstanceByKeyExample(ProcessDefinitionKey processDefinitionKey)
    {
        using var client = CamundaClient.Create();

        var result = await client.CreateProcessInstanceAsync(new ProcessInstanceCreationInstructionByKey
        {
            ProcessDefinitionKey = processDefinitionKey,
        });

        Console.WriteLine($"Process instance key: {result.ProcessInstanceKey}");
    }
    // </CreateProcessInstanceByKey>
    #endregion CreateProcessInstanceByKey

    #region GetProcessInstance

    // <GetProcessInstance>
    public static async Task GetProcessInstanceExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        var result = await client.GetProcessInstanceAsync(processInstanceKey);
        Console.WriteLine($"Process instance: {result.ProcessDefinitionId}");
    }
    // </GetProcessInstance>
    #endregion GetProcessInstance

    #region SearchProcessInstances

    // <SearchProcessInstances>
    public static async Task SearchProcessInstancesExample()
    {
        using var client = CamundaClient.Create();

        var result = await client.SearchProcessInstancesAsync(new ProcessInstanceSearchQuery());

        foreach (var instance in result.Items)
        {
            Console.WriteLine($"Process instance: {instance.ProcessInstanceKey}");
        }
    }
    // </SearchProcessInstances>
    #endregion SearchProcessInstances

    #region CancelProcessInstance

    // <CancelProcessInstance>
    public static async Task CancelProcessInstanceExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        await client.CancelProcessInstanceAsync(
            processInstanceKey,
            new CancelProcessInstanceRequest());
    }
    // </CancelProcessInstance>
    #endregion CancelProcessInstance

    #region DeleteProcessInstance

    // <DeleteProcessInstance>
    public static async Task DeleteProcessInstanceExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        await client.DeleteProcessInstanceAsync(
            processInstanceKey,
            new DeleteProcessInstanceRequest());
    }
    // </DeleteProcessInstance>
    #endregion DeleteProcessInstance

    #region MigrateProcessInstance

    // <MigrateProcessInstance>
    public static async Task MigrateProcessInstanceExample(ProcessInstanceKey processInstanceKey, ProcessDefinitionKey targetProcessDefinitionKey)
    {
        using var client = CamundaClient.Create();

        await client.MigrateProcessInstanceAsync(
            processInstanceKey,
            new ProcessInstanceMigrationInstruction
            {
                TargetProcessDefinitionKey = targetProcessDefinitionKey,
            });
    }
    // </MigrateProcessInstance>
    #endregion MigrateProcessInstance

    #region ModifyProcessInstance

    // <ModifyProcessInstance>
    public static async Task ModifyProcessInstanceExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        await client.ModifyProcessInstanceAsync(
            processInstanceKey,
            new ProcessInstanceModificationInstruction());
    }
    // </ModifyProcessInstance>
    #endregion ModifyProcessInstance

    #region GetProcessInstanceStatistics

    // <GetProcessInstanceStatistics>
    public static async Task GetProcessInstanceStatisticsExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        var result = await client.GetProcessInstanceStatisticsAsync(
            processInstanceKey);

        foreach (var stat in result.Items)
        {
            Console.WriteLine($"Element: {stat.ElementId}");
        }
    }
    // </GetProcessInstanceStatistics>
    #endregion GetProcessInstanceStatistics

    #region GetProcessInstanceSequenceFlows

    // <GetProcessInstanceSequenceFlows>
    public static async Task GetProcessInstanceSequenceFlowsExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        var result = await client.GetProcessInstanceSequenceFlowsAsync(
            processInstanceKey);

        foreach (var flow in result.Items)
        {
            Console.WriteLine($"Sequence flow: {flow}");
        }
    }
    // </GetProcessInstanceSequenceFlows>
    #endregion GetProcessInstanceSequenceFlows

    #region GetProcessInstanceCallHierarchy

    // <GetProcessInstanceCallHierarchy>
    public static async Task GetProcessInstanceCallHierarchyExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        var result = await client.GetProcessInstanceCallHierarchyAsync(
            processInstanceKey);

        Console.WriteLine($"Call hierarchy: {result}");
    }
    // </GetProcessInstanceCallHierarchy>
    #endregion GetProcessInstanceCallHierarchy

    #region SearchProcessInstanceIncidents

    // <SearchProcessInstanceIncidents>
    public static async Task SearchProcessInstanceIncidentsExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        var result = await client.SearchProcessInstanceIncidentsAsync(
            processInstanceKey,
            new IncidentSearchQuery());

        foreach (var incident in result.Items)
        {
            Console.WriteLine($"Incident: {incident.IncidentKey}");
        }
    }
    // </SearchProcessInstanceIncidents>
    #endregion SearchProcessInstanceIncidents

    #region ResolveProcessInstanceIncidents

    // <ResolveProcessInstanceIncidents>
    public static async Task ResolveProcessInstanceIncidentsExample(ProcessInstanceKey processInstanceKey)
    {
        using var client = CamundaClient.Create();

        var result = await client.ResolveProcessInstanceIncidentsAsync(
            processInstanceKey);

        Console.WriteLine($"Batch operation key: {result.BatchOperationKey}");
    }
    // </ResolveProcessInstanceIncidents>
    #endregion ResolveProcessInstanceIncidents
}

File: .sdk-repos/orchestration-cluster-api-go/examples/process_instance.go
// Process instance operations: create (by id/key), search, read, cancel, delete,
// migrate, modify, incident resolution, statistics, and batch operations.
package examples

import (
	"context"
	"fmt"

	camunda "github.com/camunda/orchestration-cluster-api-go"
	openapi "github.com/camunda/orchestration-cluster-api-go/client"
)

func createProcessInstanceByIdExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region CreateProcessInstanceById
	byID := openapi.NewProcessInstanceCreationInstructionById("order-process")
	byID.SetVariables(map[string]any{"orderId": "order-42"})

	result, err := client.CreateProcessInstance(ctx,
		openapi.ProcessInstanceCreationInstructionByIdAsProcessInstanceCreationInstruction(byID))
	if err != nil {
		return err
	}
	fmt.Printf("started instance %v\n", result.GetProcessInstanceKey())
	// endregion CreateProcessInstanceById
	return nil
}

func createProcessInstanceByKeyExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region CreateProcessInstanceByKey
	// Use a specific process definition version by its key.
	byKey := openapi.NewProcessInstanceCreationInstructionByKey(openapi.ModelString("2251799813685330"))
	byKey.SetVariables(map[string]any{"orderId": "order-42"})

	result, err := client.CreateProcessInstance(ctx,
		openapi.ProcessInstanceCreationInstructionByKeyAsProcessInstanceCreationInstruction(byKey))
	if err != nil {
		return err
	}
	fmt.Printf("started instance %v\n", result.GetProcessInstanceKey())
	// endregion CreateProcessInstanceByKey
	return nil
}

func searchProcessInstancesExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region SearchProcessInstances
	result, err := client.SearchProcessInstances(ctx, *openapi.NewProcessInstanceSearchQuery())
	if err != nil {
		return err
	}
	for _, pi := range result.GetItems() {
		fmt.Printf("%v: %v\n", pi.GetProcessInstanceKey(), pi.GetState())
	}
	// endregion SearchProcessInstances
	return nil
}

func getProcessInstanceExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region GetProcessInstance
	instance, err := client.GetProcessInstance(ctx, openapi.MustProcessInstanceKey("2251799813685340"))
	if err != nil {
		return err
	}
	fmt.Printf("state=%v definition=%q\n", instance.GetState(), instance.GetProcessDefinitionId())
	// endregion GetProcessInstance
	return nil
}

func cancelProcessInstanceExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region CancelProcessInstance
	return client.CancelProcessInstance(ctx,
		openapi.MustProcessInstanceKey("2251799813685340"),
		*openapi.NewCancelProcessInstanceRequest())
	// endregion CancelProcessInstance
}

func deleteProcessInstanceExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region DeleteProcessInstance
	return client.DeleteProcessInstance(ctx,
		openapi.MustProcessInstanceKey("2251799813685340"),
		*openapi.NewDeleteProcessInstanceRequest())
	// endregion DeleteProcessInstance
}

func getProcessInstanceCallHierarchyExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region GetProcessInstanceCallHierarchy
	hierarchy, err := client.GetProcessInstanceCallHierarchy(ctx, openapi.MustProcessInstanceKey("2251799813685340"))
	if err != nil {
		return err
	}
	for _, entry := range hierarchy {
		fmt.Printf("%v\n", entry)
	}
	// endregion GetProcessInstanceCallHierarchy
	return nil
}

func getProcessInstanceSequenceFlowsExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region GetProcessInstanceSequenceFlows
	result, err := client.GetProcessInstanceSequenceFlows(ctx, openapi.MustProcessInstanceKey("2251799813685340"))
	if err != nil {
		return err
	}
	fmt.Printf("%v\n", result)
	// endregion GetProcessInstanceSequenceFlows
	return nil
}

func getProcessInstanceStatisticsExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region GetProcessInstanceStatistics
	result, err := client.GetProcessInstanceStatistics(ctx, openapi.MustProcessInstanceKey("2251799813685340"))
	if err != nil {
		return err
	}
	fmt.Printf("%v\n", result)
	// endregion GetProcessInstanceStatistics
	return nil
}

func getProcessInstanceWaitStateStatisticsExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region GetProcessInstanceWaitStateStatistics
	result, err := client.GetProcessInstanceWaitStateStatistics(ctx, openapi.MustProcessInstanceKey("2251799813685340"))
	if err != nil {
		return err
	}
	fmt.Printf("%v\n", result)
	// endregion GetProcessInstanceWaitStateStatistics
	return nil
}

func resolveProcessInstanceIncidentsExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region ResolveProcessInstanceIncidents
	result, err := client.ResolveProcessInstanceIncidents(ctx, openapi.MustProcessInstanceKey("2251799813685340"))
	if err != nil {
		return err
	}
	fmt.Printf("%v\n", result)
	// endregion ResolveProcessInstanceIncidents
	return nil
}

func searchProcessInstanceIncidentsExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region SearchProcessInstanceIncidents
	result, err := client.SearchProcessInstanceIncidents(ctx,
		openapi.MustProcessInstanceKey("2251799813685340"),
		*openapi.NewIncidentSearchQuery())
	if err != nil {
		return err
	}
	for _, inc := range result.GetItems() {
		fmt.Printf("%v\n", inc)
	}
	// endregion SearchProcessInstanceIncidents
	return nil
}

func migrateProcessInstanceExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region MigrateProcessInstance
	instruction := openapi.NewProcessInstanceMigrationInstruction(
		openapi.ModelString("2251799813685399"),
		[]openapi.MigrateProcessInstanceMappingInstruction{
			*openapi.NewMigrateProcessInstanceMappingInstruction("review", "review-v2"),
		})

	return client.MigrateProcessInstance(ctx, openapi.MustProcessInstanceKey("2251799813685340"), *instruction)
	// endregion MigrateProcessInstance
}

func modifyProcessInstanceExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region ModifyProcessInstance
	return client.ModifyProcessInstance(ctx,
		openapi.MustProcessInstanceKey("2251799813685340"),
		*openapi.NewProcessInstanceModificationInstruction())
	// endregion ModifyProcessInstance
}

func cancelProcessInstancesBatchOperationExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region CancelProcessInstancesBatchOperation
	// Cancel every instance matching a filter in a single batch operation.
	req := openapi.NewProcessInstanceCancellationBatchOperationRequest(*openapi.NewProcessInstanceFilter())

	result, err := client.CancelProcessInstancesBatchOperation(ctx, *req)
	if err != nil {
		return err
	}
	fmt.Printf("created batch operation %v\n", result.GetBatchOperationKey())
	// endregion CancelProcessInstancesBatchOperation
	return nil
}

func deleteProcessInstancesBatchOperationExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region DeleteProcessInstancesBatchOperation
	req := openapi.NewProcessInstanceDeletionBatchOperationRequest(*openapi.NewProcessInstanceFilter())

	result, err := client.DeleteProcessInstancesBatchOperation(ctx, *req)
	if err != nil {
		return err
	}
	fmt.Printf("created batch operation %v\n", result.GetBatchOperationKey())
	// endregion DeleteProcessInstancesBatchOperation
	return nil
}

func resolveIncidentsBatchOperationExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region ResolveIncidentsBatchOperation
	req := openapi.NewProcessInstanceIncidentResolutionBatchOperationRequest(*openapi.NewProcessInstanceFilter())

	result, err := client.ResolveIncidentsBatchOperation(ctx, *req)
	if err != nil {
		return err
	}
	fmt.Printf("created batch operation %v\n", result.GetBatchOperationKey())
	// endregion ResolveIncidentsBatchOperation
	return nil
}

func migrateProcessInstancesBatchOperationExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region MigrateProcessInstancesBatchOperation
	plan := openapi.NewProcessInstanceMigrationBatchOperationPlan(
		openapi.ModelString("2251799813685399"),
		[]openapi.MigrateProcessInstanceMappingInstruction{
			*openapi.NewMigrateProcessInstanceMappingInstruction("review", "review-v2"),
		})
	req := openapi.NewProcessInstanceMigrationBatchOperationRequest(*openapi.NewProcessInstanceFilter(), *plan)

	result, err := client.MigrateProcessInstancesBatchOperation(ctx, *req)
	if err != nil {
		return err
	}
	fmt.Printf("created batch operation %v\n", result.GetBatchOperationKey())
	// endregion MigrateProcessInstancesBatchOperation
	return nil
}

func modifyProcessInstancesBatchOperationExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region ModifyProcessInstancesBatchOperation
	req := openapi.NewProcessInstanceModificationBatchOperationRequest(
		*openapi.NewProcessInstanceFilter(),
		[]openapi.ProcessInstanceModificationMoveBatchOperationInstruction{
			*openapi.NewProcessInstanceModificationMoveBatchOperationInstruction("review", "approve"),
		})

	result, err := client.ModifyProcessInstancesBatchOperation(ctx, *req)
	if err != nil {
		return err
	}
	fmt.Printf("created batch operation %v\n", result.GetBatchOperationKey())
	// endregion ModifyProcessInstancesBatchOperation
	return nil
}

func suspendProcessInstanceExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region SuspendProcessInstance
	return client.SuspendProcessInstance(ctx,
		openapi.MustProcessInstanceKey("2251799813685340"),
		*openapi.NewSuspendProcessInstanceRequest())
	// endregion SuspendProcessInstance
}

func resumeProcessInstanceExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region ResumeProcessInstance
	return client.ResumeProcessInstance(ctx,
		openapi.MustProcessInstanceKey("2251799813685340"),
		*openapi.NewResumeProcessInstanceRequest())
	// endregion ResumeProcessInstance
}

func assignProcessInstanceBusinessIdExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region AssignProcessInstanceBusinessId
	return client.AssignProcessInstanceBusinessId(ctx,
		openapi.MustProcessInstanceKey("2251799813685340"),
		*openapi.NewProcessInstanceBusinessIdAssignmentInstruction("order-42"))
	// endregion AssignProcessInstanceBusinessId
}

func suspendProcessInstancesBatchOperationExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region SuspendProcessInstancesBatchOperation
	// Suspend every instance matching a filter in a single batch operation.
	req := openapi.NewProcessInstanceSuspensionBatchOperationRequest(*openapi.NewProcessInstanceFilter())

	result, err := client.SuspendProcessInstancesBatchOperation(ctx, *req)
	if err != nil {
		return err
	}
	fmt.Printf("created batch operation %v\n", result.GetBatchOperationKey())
	// endregion SuspendProcessInstancesBatchOperation
	return nil
}

func resumeProcessInstancesBatchOperationExample(ctx context.Context, client *camunda.CamundaClient) error {
	// region ResumeProcessInstancesBatchOperation
	// Resume every previously-suspended instance matching a filter.
	req := openapi.NewProcessInstanceResumptionBatchOperationRequest(*openapi.NewProcessInstanceFilter())

	result, err := client.ResumeProcessInstancesBatchOperation(ctx, *req)
	if err != nil {
		return err
	}
	fmt.Printf("created batch operation %v\n", result.GetBatchOperationKey())
	// endregion ResumeProcessInstancesBatchOperation
	return nil
}
